mediatek: Add support for Zyxel EX5601-T0 with stock partitioning

pameruoso opened the pull request:
Hi I've been working on adding support for the Zyxel EX5601-T0 router.

Hardware
--------
SOC:   MediaTek MT7986a
CPU:   4 core cortex-a53 (2000MHz)
RAM:   1GB DDR4
FLASH: 512MB SPI-NAND (Micron xxx)
WIFI:  Mediatek MT7986 802.11ax 5 GHz 4x4 + 2.4GHZ 4x4
ETH:   MediaTek MT7531 Switch + SoC
       3 x builtin 1G phy (lan1, lan2, lan3)
       1 x MaxLinear GPY211B 2.5 N-Base-T phy5 (lan4)
       1 x MaxLinear GPY211B 2.5Gbit  xor SFP/N-Base-T phy6 (wan) 
USB:   1 x USB 3.2 Enhanced SuperSpeed port
UART:  3V3 115200 8N1 (Pinout: GND KEY RX TX VCC)
VOIP: 2 FXS ports for analog phones

Working features
--------
3 gbit lan ports
Wifi
Zyxel partitioning for coexistance with Zloader and dual boot.
WAN SFP port (only after exporting pins 57 and 10. gpiobase411)
leds
reset button
serial interface
usb port
lan ethernet 2.5 gbit port (autosense)
wan ethernet 2.5 gbit port (autosense)

Not working
--------
voip (missing drivers or proper zyxel platform software)

Swapping the wan ethernet/sfp xor port
--------
The way to swap the wan port between sfp and ethernet is the following:
export the pins 57 and 10.
Pin 57 is used to probe if an sfp is present. 
If pin 57 value is 0 it means that an sfp is present into the cage (cat /sys/class/gpio/gpio468/value)
If pin 57 value is 1 it means that no sfp is inserted into the cage.
In conclusion by default both 57 an 10 pins are by default 1, which means that the active port is the ethernet one.
After inserting an SFP pin 57 will become 0 and you have to manually change the value of pin 10 to 0 too.
This is totally scriptable of course.

Leds description
--------
All the leds are working out of the box but the leds managed by the 2 maxlinear phy (phy 5 lan, phy6 wan)
To activate the phy5 led (rj45 ethernet port led on the back of the router) you have to use mdio-tools
To activate the phy6 led (led on the front of the router for 2.5gbit link) you have to use mdio-tools
Example:
**Set lan5 led to fast blink on 2500/1000, slow blink on 10/100**
mdio mdio-bus mmd 5:30 raw 0x0001 0x33FC

**Set wan 2.5gbit led to constant on when wan is 2.5gbit**
mdio mdio-bus mmd 6:30 raw 0x0001 0x0080

Additional details regarding this router
--------
https://hack-gpon.github.io/zyxel/
https://forum.openwrt.org/t/adding-openwrt-support-for-zyxel-ex5601-t0/155914

csharper2005 reviewed (commented):
Thanks for the PR.

1. I suggest to add to the commit message at least:
- Hardware info
- Installation guide
- MAC address table
- How to recovery
2. Please check the commit title. I think it should be mediatek: Add support for some router
3. dts file names usually aren't contain the capital letters
4. Commits can be squashed into the one

csharper2005 commented on target/linux/mediatek/dts/mt7986a-zyxel-EX5601-T0.dtsi:
```diff
@@ -0,0 +1,478 @@
+
+&usb_phy {
+	status = "okay";
+};
```
Please add a newline at the end.

csharper2005 replied:
And why do you need 2 dts* files? dtsi usually contains the common blocks and definition used by the several devices.

pameruoso replied:
I m thinking about the same as Xiaomi Redmi where there is the option to compile with stock partitioning or owrt partitioning, I'm planning to replace the stock U-Boot with a clean one and sampler partitioning (one big ubi without overlay)

csharper2005 commented on target/linux/mediatek/dts/mt7986a-zyxel-ex5601-t0-stock.dts:
```diff
@@ -0,0 +1,72 @@
+			partition@100000 {
+				label = "u-boot-env";
+				reg = <0x0100000 0x0080000>;
+			};
```
I suggest make this `read-only` or add `uboot-env` config. The same for the other partitions those aren't `read-only` and aren't used by OpenWrt.

pameruoso replied:
Added U-Boot env config and read-only on other partitions

csharper2005 commented on target/linux/mediatek/dts/mt7986a-zyxel-ex5601-t0.dtsi:
```diff
@@ -0,0 +1,478 @@
+		phy5: phy@5 {
+			compatible = "ethernet-phy-ieee802.3-c45";
+			reg = <5>;
+			phy-mode = "2500base-x";
```
This line can be dropped.

pameruoso replied:
Done

csharper2005 commented on target/linux/mediatek/dts/mt7986a-zyxel-ex5601-t0.dtsi:
```diff
@@ -0,0 +1,478 @@
+		phy6: phy@6 {
+			compatible = "ethernet-phy-ieee802.3-c45";
+			reg = <6>;
+			phy-mode = "2500base-x";
```
This line can be dropped.

pameruoso replied:
Done

csharper2005 commented on target/linux/mediatek/dts/mt7986a-zyxel-ex5601-t0.dtsi:
```diff
@@ -0,0 +1,478 @@
+
+				port@6 {
+					reg = <6>;
+					label = "cpu";
```
This line can be dropped.

pameruoso replied:
Done

csharper2005 commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -229,3 +229,26 @@ ifneq ($(CONFIG_TARGET_ROOTFS_INITRAMFS),)
 endef
 TARGET_DEVICES += xiaomi_redmi-router-ax6000-ubootmod
+
+define Device/Zyxel_EX5601-T0-stock
```
Lower case would be better.

pameruoso replied:
Fixed

csharper2005 commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -229,3 +229,26 @@ ifneq ($(CONFIG_TARGET_ROOTFS_INITRAMFS),)
+
+define Device/Zyxel_EX5601-T0-stock
+  DEVICE_VENDOR := Zyxel
+  DEVICE_MODEL := EX5601-T0 (NAND)
```
Does non NAND version of the EX5601-T0 also exist?

pameruoso replied:
Only nand is possible. I removed nand and added a reference to the stock partitioning

csharper2005 commented on target/linux/mediatek/dts/mt7986a-zyxel-ex5601-t0.dtsi:
```diff
@@ -0,0 +1,478 @@
+	mdio: mdio-bus {
+		#address-cells = <1>;
+		#size-cells = <0>;
+		reset-gpios = <&pio 6 1>;
```
You can use `GPIO_ACTIVE_HIGH` instead of `0` and `GPIO_ACTIVE_LOW` instead of `1`.

pameruoso replied:
Fixed

csharper2005 commented on target/linux/mediatek/dts/mt7986a-zyxel-ex5601-t0.dtsi:
```diff
@@ -0,0 +1,478 @@
+&eth {
+	status = "okay";
+
+	gmac0: mac@0 {
```
Are the `eth0` and `eth1` mac addresses correct?

pameruoso replied:
Added Mac address table and setup in 02_networlk

csharper2005 commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -229,3 +229,26 @@ ifneq ($(CONFIG_TARGET_ROOTFS_INITRAMFS),)
+	fit lzma $$(KDIR)/image-$$(firstword $$(DEVICE_DTS)).dtb with-initrd
+  DTC_FLAGS += -@ --space 32768
+endef
+TARGET_DEVICES += Zyxel_EX5601-T0-stock
```
`Zyxel_EX5601-T0-stock` Lower case is more usual.

pameruoso replied:
Fixed

csharper2005 commented:
> filogic: add support for Zyxel EX5601-T0 router

`mediatek:` is more usual naming

> <42093926+pameruoso@users.noreply.github.com>

`Signed-off-by:`  usually contains the real Email address

MAC address table in the commit message would be nice.

Ansuel commented:
Signed off tag is still missing

Ansuel commented on target/linux/mediatek/filogic/base-files/etc/board.d/02_network:
```diff
@@ -59,6 +62,10 @@ mediatek_setup_macs()
 		wan_mac=$(mtd_get_mac_ascii Bdata ethaddr_wan)
 		label_mac=$wan_mac
 		;;
+	zyxel,ex5601-t0)
+		lan_mac=$(mtd_get_mac_binary "Factory" 0x2A)
+		wan_mac=$(mtd_get_mac_binary "Factory" 0x24)
+		;;
```
Is nvmem cell not supported by the network driver? considering Factory is a partition, data is binary... no sense to use board.d script when you can use dts directly to reference them.

Same comment for the wifi

pameruoso replied:
Can you link and example for DTS directly? I can try it with initramfs and switch to DTS... I ve used the same method of the Zyxel firmware, they map lan and wan on the board.d scripts.

csharper2005 replied:
Here is an example:
https://github.com/openwrt/openwrt/blob/master/target/linux/mediatek/dts/mt7981b-cudy-wr3000-v1.dts#L225

P.S. And `0x2a` without the caps would be more traditional for this project. :)

pameruoso replied:
While that may work for gmac0 and gmac1, what about the wifi? Without the wifi Mac fix the AP 2.4 takes the right Mac address but the wifi 5 is taking a random one, hence the need to fix it in wifi2.4 Mac+1

pameruoso replied:
Team, I did like you suggested, I removed the board.d configs and added mac address configurations into gmacs. Everything works as expected but the Wifi. Like I mentioned in my previous comment, the two wifi have 2 different mac addresses where wifi 2.4 ghz is 0x0004 from factory and wifi 5.ghz is 0x0004 +1. With the current commit wifi 2.4 and 5 have the same mac-address which is not correct. What should I use to fix the wifi 5ghz mac address?

csharper2005 replied:
Of course you can use `base-files/etc/hotplug.d/ieee80211/11_fix_wifi_mac` for fixing 5ghz MAC.

pameruoso replied:
Done

pameruoso committed:
mediatek: add support for Zyxel EX5601-T0 router

Zyxel EX5601-T0 specifics
--------------
The operator specific firmware running on the Zyxel branded
EX5601-T0 includes  U-Boot modifications affecting the OpenWrt
installation.

Partition Table
| dev  | size     | erasesize | name          |
| ---- | -------- | --------- | ------------- |
| mtd0 | 20000000 | 00040000  | "spi0.1"      |
| mtd1 | 00100000 | 00040000  | "BL2"         |
| mtd2 | 00080000 | 00040000  | "u-boot-env"  |
| mtd3 | 00200000 | 00040000  | "Factory"     |
| mtd4 | 001c0000 | 00040000  | "FIP"         |
| mtd5 | 00040000 | 00040000  | "zloader"     |
| mtd6 | 04000000 | 00040000  | "ubi"         |
| mtd7 | 04000000 | 00040000  | "ubi2"        |
| mtd8 | 15a80000 | 00040000  | "zyubi"       |

The router boots BL2 which than loads FIP (u-boot).
U-boot has hardcoded a command to always launch Zloader "mtd read zloader 0x46000000" and than "bootm". Bootargs are deactivated.
Zloader is the zyxel booloader which allow to dual-boot ubi or ubi2, by default access to zloader is blocked.
Too zloader checks that the firmware contains a particolar file called zyfwinfo.
Additional details regarding Zloader can be found here:
https://hack-gpon.github.io/zyxel/
https://forum.openwrt.org/t/adding-openwrt-support-for-zyxel-ex5601-t0/155914

Hardware
--------
SOC: MediaTek MT7986a
CPU: 4 core cortex-a53 (2000MHz)
RAM: 1GB DDR4
FLASH: 512MB SPI-NAND (Micron xxx)
WIFI: Wifi6 Mediatek MT7976 802.11ax 5 GHz 4x4 + 2.4GHZ 4x4
ETH: MediaTek MT7531 Switch + SoC
3 x builtin 1G phy (lan1, lan2, lan3)
1 x MaxLinear GPY211B 2.5 N-Base-T phy5 (lan4)
1 x MaxLinear GPY211B 2.5Gbit xor SFP/N-Base-T phy6 (wan)
USB: 1 x USB 3.2 Enhanced SuperSpeed port
UART: 3V3 115200 8N1 (Pinout: GND KEY RX TX VCC)
VOIP: 2 FXS ports for analog phones

MAC Address Table
-----------------
eth0/lan    Factory 0x002a
eth1/wan    Factory 0x0024
wifi 2.4Ghz Factory 0x0004
wifi 5Ghz   Factory 0x0004 + 1

Serial console (UART)
---------------------
+-------+-------+-------+-------+-------+
| +3.3V |  RX   |  TX   |  KEY  |  GND  |
+---+---+-------+-------+-------+-------+
    |
    +--- Don't connect

Installation
------------
Keep in mind that openwrt can only run on the UBI partition, the openwrt firmware is not able to understand the zloader bootargs.
The procedure allows restoring the UBI partition with the Zyxel firmware and retains all the OEM functionalities.

1. Unlock Zloader (this will allow to swap manually between partitions UBI and UBI2):
- Attach a usb-ttl adapter to your computer and boot the router.
- While the router is booting at some point you will read the following: `Please press Enter to activate this console.`
- As soon as you read that press enter, type root and than press enter again (just do it, don't care about the logs scrolling).
- Most likely the router is still printing the boot log, leave it boot until it stops.
- If everything went ok you should have full root access "root@EX5601-T0:/#".
- Type the following command and press enter: "fw_setenv EngDebugFlag 0x1".
- Reboot the router.
- As soon as you read `Hit any key to stop autoboot:` press Enter.
- If everything went ok you should have the following prompt: "ZHAL>".
- You have successfully unlocked zloader access, this procedure must be done only once.

2. Check the current active partition:
- Boot the router and repeat the steps above to gain root access.
- Type the following command to check the current active image: "cat /proc/cmdline".
- If `rootubi=ubi` it means that the active partition is `mtd6`
- If `rootubi=ubi2` it means that the active partition is `mtd7`
- As mentioned earlier we need to flash openwrt into ubi/mtd6 and never overwrite ubi2/mtd7 to be able to fully roll-back.
- To activate and boot from mtd7 (ubi2) enter into ZHAL> command prompt and type the following commands:
atbt 1  # unlock write
atsw    # swap boot partition
atsr    # reboot the router
- After rebooting check again with "cat /proc/cmdline" that you are correctly booting from mtd7/ubi2
- If yes proceed with the installation guide. If not probably you don't have a firmware into ubi2 or you did something wrong.

3. Flashing:
- Download the sysupgrade file for the router from openwrt, than we need to add the zyfwinfo file into the sysupgrade tar.
Zloader only checks for the magic (which is a fixed value 'EXYZ') and the crc of the file itself (256bytes).
I created a script to create a valid zyfwinfo file but you can use anything that does exactly the same:
https://raw.githubusercontent.com/pameruoso/OpenWRT-Zyxel-EX5601-T0/main/gen_zyfwinfo.sh
- Add the zyfwinfo file into the sysupgrade tar.
- Enter via telnet or ssh into the router with admin credentials
- Enter the following commands to disable the firmware and model checks
"zycli fwidcheck off" and "zycli modelcheck off"
- Open the router web interface and in the update firmware page select the "restore default settings option"
- Select the sysupgrade file and click on upload.
- The router will flash and reboot itself into openwrt from UBI

4. Restoring and going back to Zyxel firmware.
- Use the ZHAL> command line to manually swap the boot parition to UBI2 with the following:
atbt 1  # unlock write
atsw    # swap boot partition
atsr    # reboot the router
- You will boot again the Zyxel firmware you have into UBI2 and you can flash the zyxel firmware to overwrite the UBI partition and openwrt.

Working features
----------------
3 gbit lan ports
Wifi
Zyxel partitioning for coexistance with Zloader and dual boot.
WAN SFP port (only after exporting pins 57 and 10. gpiobase411)
leds
reset button
serial interface
usb port
lan ethernet 2.5 gbit port (autosense)
wan ethernet 2.5 gbit port (autosense)

Not working
----------------
voip (missing drivers or proper zyxel platform software)

Swapping the wan ethernet/sfp xor port
----------------
The way to swap the wan port between sfp and ethernet is the following:
export the pins 57 and 10.
Pin 57 is used to probe if an sfp is present.
If pin 57 value is 0 it means that an sfp is present into the cage (cat /sys/class/gpio/gpio468/value).
If pin 57 value is 1 it means that no sfp is inserted into the cage.
In conclusion by default both 57 an 10 pins are by default 1, which means that the active port is the ethernet one.
After inserting an SFP pin 57 will become 0 and you have to manually change the value of pin 10 to 0 too.
This is totally scriptable of course.

Leds description
------------
All the leds are working out of the box but the leds managed by the 2 maxlinear phy (phy 5 lan, phy6 wan).
To activate the phy5 led (rj45 ethernet port led on the back of the router) you have to use mdio-tools.
To activate the phy6 led (led on the front of the router for 2.5gbit link) you have to use mdio-tools.
Example:
Set lan5 led to fast blink on 2500/1000, slow blink on 10/100:
mdio mdio-bus mmd 5:30 raw 0x0001 0x33FC

Set wan 2.5gbit led to constant on when wan is 2.5gbit:
mdio mdio-bus mmd 6:30 raw 0x0001 0x0080

Signed-off-by: Pietro Ameruoso <p.ameruoso@live.it>

csharper2005 commented on target/linux/mediatek/dts/mt7986a-zyxel-ex5601-t0-stock.dts:
```diff
@@ -0,0 +1,563 @@
+		reg = <0x0024 0x6>;
+	};
+
+	macaddr_factory_0004: macaddr@0004 {
```
Would you like to sort the blocks alphabetically? 4, 24, 2a.

pameruoso replied:
Done

Ansuel commented on target/linux/mediatek/dts/mt7986a-zyxel-ex5601-t0-stock.dts:
```diff
@@ -0,0 +1,563 @@
+		};
+	};
+};
+
```
extra space?

pameruoso commented on target/linux/mediatek/dts/mt7986a-zyxel-ex5601-t0-stock.dts:
```diff
@@ -0,0 +1,563 @@
+	nvmem-cells = <&macaddr_factory_0004>;
+	nvmem-cell-names = "mac-address";
+};
+
```
Sorry again fixed

1715173329 commented:
the dts file needs dos2unix imo
